sync2/rangesync/wire_conduit.go

```diff
@@ -0,0 +1,222 @@
+)
+
+const (
+	writeQueueSize = 10000
+)
+
+var ErrLimitExceeded = errors.New("sync traffic/message limit exceeded")
+
+// ConduitOption specifies an option for a message conduit.
+type ConduitOption func(c *wireConduit)
+
+// WithTrafficLimit sets a limit on the total number of bytes sent and received.
+// Zero or negative values disable the limit.
+func WithTrafficLimit(limit int) ConduitOption {
+	return func(c *wireConduit) {
+		c.trafficLimit = limit
+	}
+}
+
+// WithMessageLimit sets a limit on the total number of messages sent and received.
+// Zero or negative values disable the limit.
+func WithMessageLimit(limit int) ConduitOption {
+	return func(c *wireConduit) {
+		c.messageLimit = limit
+	}
+}
+
+type wireConduit struct {
+	stream       io.ReadWriter
+	eg           errgroup.Group
+	sendCh       chan SyncMessage
+	stopCh       chan struct{}
+	nBytesSent   atomic.Int64
+	nBytesRecv   atomic.Int64
+	nMsgsSent    atomic.Int64
+	nMsgsRecv    atomic.Int64
+	trafficLimit int
+	messageLimit int
+}
+
+var _ Conduit = &wireConduit{}
+
+func startWireConduit(ctx context.Context, s io.ReadWriter, opts ...ConduitOption) *wireConduit {
+	c := &wireConduit{
+		stream: s,
+		sendCh: make(chan SyncMessage, writeQueueSize),
+		stopCh: make(chan struct{}),
+	}
+	for _, opt := range opts {
+		opt(c)
+	}
+	c.eg.Go(func() error {
+		defer close(c.stopCh)
+		for {
+			select {
+			case <-ctx.Done():
+				c.closeStream()
+				return ctx.Err()
+			case m, ok := <-c.sendCh:
+				if !ok {
+					return nil
+				}
+				n, err := writeMessage(c.stream, m)
+				c.nBytesSent.Add(int64(n))
+				c.nMsgsSent.Add(1)
+				if err == nil {
+					err = c.checkLimits()
+				}
+				if err != nil {
+					c.closeStream()
+					return err
+				}
+			}
+		}
+	})
+	return c
+}
+
+func (c *wireConduit) closeStream() {
+	if closer, ok := c.stream.(io.Closer); ok {
+		closer.Close()
+	}
+}
+
+func (c *wireConduit) stop() {
+	if c.stream == nil {
+		return
+	}
+	// if there was in error, there's no point in waiting for the send
+	// goroutine to finish, so we interrupt it by closing the stream
+	c.closeStream()
+	c.end()
+}
+
+func (c *wireConduit) end() {
+	if c.stream == nil {
+		return
+	}
+	close(c.sendCh)
+	c.eg.Wait()
+	c.stream = nil
+}
+
+func (c *wireConduit) checkLimits() error {
+	if c.trafficLimit > 0 && c.bytesSent()+c.bytesReceived() > c.trafficLimit {
+		return ErrLimitExceeded
+	}
+	if c.messageLimit > 0 && c.messagesSent()+c.messagesReceived() > c.trafficLimit {
+		return ErrLimitExceeded
+	}
+	return nil
+}
+
+func (c *wireConduit) NextMessage() (SyncMessage, error) {
+	msg, n, err := c.nextMessage()
+	c.nBytesRecv.Add(int64(n))
+	c.nMsgsRecv.Add(1)
+	if err != nil {
+		return nil, fmt.Errorf("receive message: %w", err)
+	}
+	if err = c.checkLimits(); err != nil {
+		return nil, err
+	}
+	return msg, nil
+}
+
+func (c *wireConduit) nextMessage() (SyncMessage, int, error) {
+	var b [1]byte
+	if n, err := io.ReadFull(c.stream, b[:]); err != nil {
```

Comment:
What is there a reason for passing there `b[:]` instead of just `b`?

Reply:
`io.ReadFull` expects a slice as second parameter, but `b` is an array. `[:]` converts an array to a slice 🙂

Reply:
`b` is an 1-element array, but we need to pass a slice here

Reply:
Ack, I get it. I was used to other form which does not require making a subslice, like:
```go
b := make([]byte, 1)
... io.ReadFull(c.stream, b)
```